File: cpp/0014最长公共前缀.cpp
#include<iostream>
#include<vector>
using namespace std;

// 暴力
string longestCommonPrefix1(vector<string> &strs)
{
    int size = strs.size();
    if (!size)
        return "";
    
    for (int i = 0;; ++i)
    {
        char ch = strs[0][i];
        for (int j = 1; j < size; ++j)
        {
            if(ch!=strs[j][i])
            {
                return strs[0].substr(0, i);
            }
        }

        if(strs[0][i]=='\0')
            return strs[0];
    }
}


// 分治
string _lcp(vector<string> &strs,int start, int end)
{
    if (start == end)
        return strs[start];
    if (start > end)
        return "";
    int mid = (start + end) / 2;
    string llcp = _lcp(strs, start, mid);
    string rlcp = _lcp(strs, mid+1, end);
    for (int i = 0;; ++i)
    {
        if(llcp[i]!=rlcp[i])
            return llcp.substr(0, i);
        
        if (llcp[i] == '\0')
            return llcp;
    }

}
string longestCommonPrefix2(vector<string> &strs)
{
    if(!strs.size())
        return "";
    return _lcp(strs, 0, strs.size()-1);
}

int main()
{
    vector<string> strs = {"flower", "flow", "flight"};

    cout<< longestCommonPrefix2(strs);
}

File: python/main/0014最长公共前缀.py


strs = ["ads","a"]

# 暴力
def longestCommonPrefix1( strs: list) -> str:
    
    for i in range(200):
        try:
            ch=strs[0][i]
            for j in strs:
                if(ch!=j[i]):
                    return j[:i] 
        except:
            return strs[0][:i]
            
# 分治
def longestCommonPrefix2( strs: list) -> str:
    
    def _lcp(start,end):
        if(start==end):
            return strs[start]
        mid=(start+end)//2
        
        llcp=_lcp(start,mid)
        rlcp=_lcp(mid+1,end)

        min_len=min(len(llcp),len(rlcp))
        for i in range(min_len):
            if(llcp[i]!=rlcp[i]):
                return llcp[:i]
        return llcp[:min_len]

    return "" if(not strs) else _lcp(0,len(strs)-1)




    

print(longestCommonPrefix2(strs))